daemon/src/db/time_to_first_position.rs

```diff
@@ -0,0 +1,262 @@
+use anyhow::Result;
+use model::Identity;
+use sqlx::pool::PoolConnection;
+use sqlx::Sqlite;
+use time::OffsetDateTime;
+
+use crate::db::Connection;
+
+impl Connection {
+    /// Record the time at which we hear about a `taker_id` for the
+    /// first time.
+    ///
+    /// If we have already heard about the `taker_id`, we keep the
+    /// original `timestamp`.
+    pub async fn try_insert_first_seen(
+        &self,
+        taker_id: Identity,
+        timestamp: OffsetDateTime,
+    ) -> Result<()> {
+        let mut conn = self.inner.acquire().await?;
+
+        if !have_seen_before(&mut conn, taker_id).await? {
+            insert_first_seen(&mut conn, taker_id, timestamp).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Record the time at which we hear about a taker with `taker_id`
+    /// opening a position for the first time.
+    ///
+    /// If we have already heard about the taker opening a position
+    /// before, we keep the original `timestamp`.
+    pub async fn try_insert_first_position(
+        &self,
+        taker_id: Identity,
+        timestamp: OffsetDateTime,
+    ) -> Result<()> {
+        let mut conn = self.inner.acquire().await?;
+
+        if !have_opened_a_position_before(&mut conn, taker_id).await? {
+            insert_first_position(&mut conn, taker_id, timestamp).await?;
+        }
+
+        Ok(())
+    }
+}
+
+async fn have_seen_before(conn: &mut PoolConnection<Sqlite>, taker_id: Identity) -> Result<bool> {
+    let first_seen_timestamp = load_first_seen_timestamp(conn, taker_id).await?;
+
+    Ok(first_seen_timestamp.is_some())
+}
+
+async fn load_first_seen_timestamp(
+    conn: &mut PoolConnection<Sqlite>,
+    taker_id: Identity,
+) -> Result<Option<OffsetDateTime>> {
+    let row = sqlx::query!(
+        r#"
+        SELECT
+            time_to_first_position.first_seen_timestamp
+        FROM
+            time_to_first_position
+        WHERE
+            time_to_first_position.taker_id = $1
+        "#,
+        taker_id
+    )
+    .fetch_optional(&mut *conn)
+    .await?;
+
+    let timestamp = match row {
+        None => return Ok(None),
+        Some(row) => row.first_seen_timestamp,
+    };
+
+    let timestamp = match timestamp {
+        None => return Ok(None),
+        Some(timestamp) => timestamp,
+    };
+
+    let timestamp = OffsetDateTime::from_unix_timestamp(timestamp)?;
```

Comment:
Might be a usecase for `transpose` https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose :) 

Maybe like this: 

```
timestamp
            .map(|row| OffsetDateTime::from_unix_timestamp(row.first_seen_timestamp))
            .transpose()?
```

Reply:
I tried it originally, but it was not very readable at all haha, because it's actually a `Result<Option<Option<OffsetDateTime>>` and I don't think you can pattern match much since `sqlx::query!` outputs "an ad-hoc anonymous struct type".

But do give it a go and see if you can produce something better!

Reply:
Pushed one commit just for fun :) https://github.com/itchysats/itchysats/pull/1948/commits/a876482ac26d4dd106dec23a8a61e3b7b96059d2